File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.cpp
class Solution {
public:
    bool check(int pos, vector<int> &nums) {
        int cnt = 0;
        for(int i = 0; i < nums.size(); i++) {
            cnt += (nums[i] <= pos);
        }
        return cnt > pos;
    }
    int findDuplicate(vector<int>& nums) {
        int n = nums.size();
        int lo = 1, hi = n - 1;
        int ans = 0;
        
        while(lo <= hi) {
            int mid = (lo + hi) / 2;
            if(check(mid, nums)) {
                ans = mid;
                hi = mid - 1;
            }
            else {
                lo = mid + 1;
            }
        }
        return ans;
    }
};

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.go
func findDuplicate(nums []int) int {
    var n int = len(nums)
    for i := 0; i < n; i++ {
        x := int(math.Abs(float64(nums[i])))
        if nums[x] < 0 {
            return x
        }
        nums[x] *= -1
    }
    return 1;
}